File: src/lib/api/workspace.ts
import { access, readdir } from 'node:fs/promises';
import { constants } from 'node:fs';
import path from 'node:path';
import { invoke } from '@tauri-apps/api/core';
import {
  parseWorkspaceSnapshot,
  type WorkspaceSnapshot,
  type WorkspaceSource,
} from '../../features/workspace-discovery/model';

export interface DiscoverWorkspaceInput {
  cwd?: string;
  homeDir?: string;
  manualPath?: string;
}

const REQUIRED_KEY_FILES = ['AGENTS.md', 'memory-bank/prd.md', 'package.json', 'src-tauri/tauri.conf.json'] as const;

async function pathExists(value: string): Promise<boolean> {
  try {
    await access(value, constants.F_OK);
    return true;
  } catch {
    return false;
  }
}

function defaultWorkspaceCandidates(cwd?: string, homeDir?: string): string[] {
  const candidates: string[] = [];
  if (cwd) {
    candidates.push(cwd);
  }
  if (homeDir) {
    candidates.push(path.join(homeDir, '.claw-workbench', 'workspace'));
    candidates.push(path.join(homeDir, '.openclaw', 'workspace'));
  }
  return candidates;
}

async function summarizeSkills(resolvedPath: string) {
  const skillsRoot = path.join(resolvedPath, 'skills');
  if (!(await pathExists(skillsRoot))) {
    return { total_skills: 0, total_files: 0, skill_names: [] as string[] };
  }

  const skillEntries = await readdir(skillsRoot, { withFileTypes: true });
  const directories = skillEntries.filter((entry) => entry.isDirectory()).map((entry) => entry.name).sort();

  let totalFiles = 0;
  for (const skillName of directories) {
    const skillFolder = path.join(skillsRoot, skillName);
    const files = await readdir(skillFolder, { withFileTypes: true });
    totalFiles += files.filter((entry) => entry.isFile()).length;
  }

  return {
    total_skills: directories.length,
    total_files: totalFiles,
    skill_names: directories,
  };
}

export async function discoverWorkspaceFromFilesystem(input: DiscoverWorkspaceInput = {}): Promise<WorkspaceSnapshot> {
  const candidates = defaultWorkspaceCandidates(input.cwd, input.homeDir);

  let resolvedPath: string | undefined;
  let source: WorkspaceSource = 'default';

  for (const candidate of candidates) {
    if (await pathExists(candidate)) {
      resolvedPath = candidate;
      source = 'default';
      break;
    }
  }

  if (!resolvedPath && input.manualPath && (await pathExists(input.manualPath))) {
    resolvedPath = input.manualPath;
    source = 'manual';
  }

  if (!resolvedPath) {
    throw new Error('No workspace found from default candidates or manual fallback path');
  }

  const key_files = await Promise.all(
    REQUIRED_KEY_FILES.map(async (relativePath) => ({
      relative_path: relativePath,
      exists: await pathExists(path.join(resolvedPath, relativePath)),
    })),
  );

  const skills = await summarizeSkills(resolvedPath);

  return parseWorkspaceSnapshot({
    resolved_path: resolvedPath,
    source,
    key_files,
    skills,
  });
}

export async function discoverWorkspaceViaTauri(input: DiscoverWorkspaceInput = {}): Promise<WorkspaceSnapshot> {
  const response = await invoke('discover_workspace', {
    manualPath: input.manualPath ?? null,
    cwd: input.cwd ?? null,
    homeDir: input.homeDir ?? null,
  });
  return parseWorkspaceSnapshot(response);
}

File: src-tauri/src/workspace.rs
use serde::Serialize;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Serialize)]
pub struct WorkspaceKeyFile {
    relative_path: String,
    exists: bool,
}

#[derive(Serialize)]
pub struct WorkspaceSkillSummary {
    total_skills: usize,
    total_files: usize,
    skill_names: Vec<String>,
}

#[derive(Serialize)]
pub struct WorkspaceSnapshot {
    resolved_path: String,
    source: String,
    key_files: Vec<WorkspaceKeyFile>,
    skills: WorkspaceSkillSummary,
}

fn default_candidates(cwd: Option<&str>, home_dir: Option<&str>) -> Vec<PathBuf> {
    let mut candidates = Vec::new();

    if let Some(cwd_value) = cwd {
        candidates.push(PathBuf::from(cwd_value));
    }

    if let Some(home) = home_dir {
        candidates.push(Path::new(home).join(".claw-workbench").join("workspace"));
        candidates.push(Path::new(home).join(".openclaw").join("workspace"));
    }

    candidates
}

fn summarize_skills(workspace_path: &Path) -> WorkspaceSkillSummary {
    let skills_root = workspace_path.join("skills");
    if !skills_root.exists() {
        return WorkspaceSkillSummary {
            total_skills: 0,
            total_files: 0,
            skill_names: vec![],
        };
    }

    let mut skill_names: Vec<String> = vec![];
    let mut total_files = 0;

    if let Ok(entries) = fs::read_dir(skills_root) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_dir() {
                if let Some(name) = path.file_name().and_then(|s| s.to_str()) {
                    skill_names.push(name.to_string());
                }
                if let Ok(skill_files) = fs::read_dir(path) {
                    total_files += skill_files.flatten().filter(|file| file.path().is_file()).count();
                }
            }
        }
    }

    skill_names.sort();

    WorkspaceSkillSummary {
        total_skills: skill_names.len(),
        total_files,
        skill_names,
    }
}

#[tauri::command]
pub fn discover_workspace(
    manual_path: Option<String>,
    cwd: Option<String>,
    home_dir: Option<String>,
) -> Result<WorkspaceSnapshot, String> {
    let mut resolved: Option<PathBuf> = None;
    let mut source = "default".to_string();

    for candidate in default_candidates(cwd.as_deref(), home_dir.as_deref()) {
        if candidate.exists() {
            resolved = Some(candidate);
            source = "default".to_string();
            break;
        }
    }

    if resolved.is_none() {
        if let Some(manual) = manual_path {
            let path = PathBuf::from(manual);
            if path.exists() {
                resolved = Some(path);
                source = "manual".to_string();
            }
        }
    }

    let resolved_path = resolved.ok_or_else(|| {
        "No workspace found from default candidates or manual fallback path".to_string()
    })?;

    let key_file_names = vec![
        "AGENTS.md",
        "memory-bank/prd.md",
        "package.json",
        "src-tauri/tauri.conf.json",
    ];

    let key_files = key_file_names
        .iter()
        .map(|relative| WorkspaceKeyFile {
            relative_path: relative.to_string(),
            exists: resolved_path.join(relative).exists(),
        })
        .collect::<Vec<_>>();

    let skills = summarize_skills(&resolved_path);

    Ok(WorkspaceSnapshot {
        resolved_path: resolved_path.to_string_lossy().to_string(),
        source,
        key_files,
        skills,
    })
}
